src/shogun/lib/any.h

```diff
@@ -1164,26 +1170,67 @@ namespace shogun {
 
 	private:
 		typedef std::type_index TypeIndex;
-		typedef std::function<bool(void*, void*)> CastingFunction;
+
+		typedef std::function<void(void*, void*)> CastingFunction;
 		typedef std::map<std::pair<TypeIndex, TypeIndex>, CastingFunction> CastingRegistry;
 		static CastingRegistry casting_registry;
+
+		typedef std::function<void(void*, void*)> VisitorFunction;
+		typedef std::map<std::pair<TypeIndex, TypeIndex>, VisitorFunction> VisitorRegistry;
+		static VisitorRegistry visitor_registry;
 	};
 
 	bool operator==(const Any& lhs, const Any& rhs);
 
 	bool operator!=(const Any& lhs, const Any& rhs);
 
+	template <class From, class To>
+	void Any::register_caster(std::function<To(From)> caster)
+	{
+		const auto key = std::make_pair(std::type_index{typeid(From)}, std::type_index{typeid(To)});
+
+		if (casting_registry.count(key))
+		{
+			return;
+		}
+
+		casting_registry[key] = [caster] (void* src, void* dst)
+		{
+			auto typed_src = static_cast<From*>(src);
+			auto typed_dst = static_cast<To*>(dst);
```

Comment:
out of curiosity how come you switched from reinterpret_cast to static_cast?

Reply:
We don't actually need reinterpret so I used more narrow one.